include/iterlib/Item.h

```diff
@@ -0,0 +1,229 @@
+static const id_t INVALID_ID = -1;
+
+class Item : public dynamic {
+ public:
+  using dynamic::operator=;
+  using dynamic::is_of;
+  using dynamic::get;
+  using dynamic::getRef;
+  using dynamic::getNonConstRef;
+
+  static const Item kEmptyItem;
+  Item() : dynamic() {}
+  Item(const dynamic& other) : dynamic(other) {}
+
+  Item(const Item& other) = default;
+  Item& operator=(const Item& other) = default;
+  Item& operator=(Item&& other) = default;
+
+  bool operator==(const Item& other) const {
+    return dynamic::operator==(static_cast<const dynamic&>(other));
+  }
+
+  id_t id() const {
+    try {
+      return at(QueryConstants::kIdKey).get<int64_t>();
+    } catch (const std::exception& ex) {
+      LOG_EVERY_N(WARNING, 1000) << ex.what();
+    }
+  }
+
+  int64_t ts() const {
+    try {
+      return at(QueryConstants::kTimeKey).get<int64_t>();
+    } catch (const std::exception& ex) {
+      LOG_EVERY_N(WARNING, 1000) << ex.what();
+    }
+  }
+
+  const dynamic& value() const {
+    return *this;
+  }
+};
+
+// Provides O(1) access to id() and ts(). Otherwise identical to
+// Item
+class ItemOptimized : public Item {
+ public:
+  using dynamic::operator=;
+  using dynamic::is_of;
+  using dynamic::get;
+  using dynamic::getRef;
+  using dynamic::getNonConstRef;
+
+  ItemOptimized()
+    : Item()
+    , id_(0)
+    , ts_(0) {}
+
+  ItemOptimized(id_t id,
+            int64_t ts,
+            dynamic value = dynamic())
+      : Item(std::move(value))
+      , id_(id)
+      , ts_(ts) {
+   }
+
+  ItemOptimized(const ItemOptimized& other) = default;
+  ItemOptimized& operator=(const ItemOptimized& other) = default;
+  ItemOptimized& operator=(ItemOptimized&& other) = default;
+
+  void reset() {
+    id_ = 0;
+    ts_ = 0;
+    *this = dynamic();
+  }
+
+  id_t id() const {
+    return id_;
+  }
+
+  void setId(id_t id) {
+    id_ = id;
+  }
+
+  void setTs(int64_t ts) {
+    ts_ = ts;
+  }
+
+  int64_t ts() const {
+    return ts_;
+  }
+
+  void syncIdTs() {
+    try {
+      id_ = at(QueryConstants::kIdKey).get<int64_t>();
+      ts_ = at(QueryConstants::kTimeKey).get<int64_t>();
+    } catch (const std::exception& ex) {
+      LOG_EVERY_N(WARNING, 1000) << ex.what();
+    }
+  }
+
+  // This depends on value().which(). For eg: vector_pair_t can't
+  // be modified.
+  void syncMap() {
+    try {
+      (*this)[QueryConstants::kIdKey] = dynamic(int64_t(id_));
+      (*this)[QueryConstants::kTimeKey] = dynamic(int64_t(ts_));
+    } catch (const std::exception& ex) {
+      LOG_EVERY_N(WARNING, 1000) << ex.what();
+    }
+  }
+
+  void sync() {
+    // This is a heuristic. The actual sync depends on what was
+    // modified. Use syncIdTs/syncMap directly depending on the
+    // context.
+    if (id_ == 0) {
+      syncIdTs();
+    } else {
+      syncMap();
+    }
+  }
+
+  const dynamic& value() const {
+    return *this;
+  }
+
+  bool operator<(const ItemOptimized& other) const {
+    return std::tie(value(), ts_, id_)
```

Comment:
Provide a `private` member function (e.g., `tied`). Then this function will look like:

```
return tied() < other.tied();
```

DRYer, less risk of ordering bugs.